File: ObjC_Rc4/ObjC_Rc4/Rc4.m
#import "Rc4.h"

@implementation Rc4

+ (void)byteArr:(UInt8 *)array forString:(NSString *)string {
  for (int p = 0; p < (int)string.length; p++) {
    array[p] = [string characterAtIndex:p];
  }
}

+ (void)decrypt:(UInt8 *)data cipher:(UInt8 *)cipher cipherLength:(int)cipherLength key:(UInt8 *)key keyLength:(int)keyLength {
  [Rc4 encrypt:data data:cipher dataLength:cipherLength key:key keyLength:keyLength];
}

+ (void)encrypt:(UInt8 *)cipher data:(UInt8 *)data dataLength:(int)dataLength key:(UInt8 *)key keyLength:(int)keyLength {
  UInt8 iS[256];
  [Rc4 keyScheduler:iS key:key keyLength:keyLength];
  [Rc4 createCipher:cipher data:data dataLength:dataLength iS:iS];
}

+ (void)keyScheduler:(UInt8 *)iS key:(UInt8 *)key keyLength:(int)keyLength {
  UInt8 iK[256];
  int i;
  for (i = 0; i < 256; i++) {
    iS[i] = i;
    iK[i] = key[i % keyLength];
  }
  
  int j = 0;
  for (i = 0; i < 256; i++) {
    int is = iS[i];
    int ik = iK[i];
    j = (j + is + ik) % 256;
    [Rc4 swap:iS forIndex:i andIndex:j];
  }
}

+ (void)swap:(UInt8 *)a forIndex:(int)firstIndex andIndex:(int)secondIndex {
  UInt8 temp = a[firstIndex];
  a[firstIndex] = a[secondIndex];
  a[secondIndex] = temp;
}

+ (void)createCipher:(UInt8 *)cipher data:(UInt8 *)data dataLength:(int)dataLength iS:(UInt8 *)iS {
  int i = 0;
  int j = 0;
  for (int x = 0; x < dataLength; x++) {
    i = (i + 1) % 256;
    int is_i = iS[i];
    
    j = (j + is_i) % 256;
    int is_j = iS[j];

    [Rc4 swap:iS forIndex:i andIndex:j];

    int k = iS[(is_i + is_j) % 256];
    cipher[x] = data[x] ^ k;
  }
}

@end

File: Swift_Rc4/Swift_Rc4/Rc4.swift
import Foundation

class Rc4 {
  func byteArr(_ str: String) -> [UInt8] {
    var result = [UInt8]()
    for strByte in str.utf8 {
      result.append(strByte)
    }
    
    return result
  }
  
  // decrypt really just calls encrypt. Amazing right?
  func decrypt(_ data: [UInt8], key: [UInt8]) -> [UInt8] {
    return encrypt(data, key: key)
  }

  func encrypt(_ data: [UInt8], key: [UInt8]) -> [UInt8] {
    let swappedArray = keyScheduler(key)
    return cipherArray(data, swappedArr: swappedArray)
  }
  
  func keyScheduler(_ key: [UInt8]) -> [UInt8] {
    let keySwap = initialKeyArrays(key)
    var iS = keySwap.iS
    let iK = keySwap.iK
    var j = 0
    for i in 0..<256 {
      let swapValue: UInt8 = iS[i]
      j = (j + Int(iS[i]) + Int(iK[i])) % 256
      iS[i] = iS[j]
      iS[j] = swapValue
    }
    
    return iS
  }
  
  func initialKeyArrays(_ keyArr: [UInt8]) -> (iS: [UInt8], iK: [UInt8]) {
    var iS = [UInt8]()
    var iK = [UInt8]()
    let keyLength = keyArr.count
    for i in 0..<256 {
      iS.append(UInt8(i))
      iK.append(keyArr[i % keyLength])
    }
        
    return (iS, iK)
  }
  
  func cipherArray(_ dataArr: [UInt8], swappedArr: [UInt8]) -> [UInt8] {
    var iPos = 0
    var jPos = 0
    var cipher = dataArr
    var swapped = swappedArr
    for cipherPos in 0..<dataArr.count {
      iPos = (iPos + 1) % 256
      let is_i = Int(swapped[iPos])
      
      jPos = (jPos + is_i) % 256
      let is_j = Int(swapped[jPos])

      swapped[iPos] = UInt8(is_j)
      swapped[jPos] = UInt8(is_i)
      let swap = swapped[(is_i + is_j) % 256]
      
      let ch = dataArr[cipherPos]
      cipher[cipherPos] = ch ^ swap
    }
    
    return cipher
  }
}
